perception_dataset/t4_dataset/data_interpolator.py

```diff
@@ -0,0 +1,614 @@
+
+        all_ego_poses = self.interpolate_ego_pose(nusc)
+        self.logger.info("Finish interpolating ego pose")
+
+        all_samples = self.interpolate_sample(nusc)
+        self.logger.info("Finish interpolating sample")
+
+        all_sample_data = self.interpolate_sample_data(nusc, all_ego_poses, all_samples)
+        self.logger.info("Finish interpolating sample data")
+
+        all_sample_anns = self.interpolate_sample_annotation(nusc, all_samples)
+        self.logger.info("Finish interpolating sample annotation")
+
+        all_instances = self.update_instance_record(nusc, all_sample_anns)
+        self.logger.info("Finish updating instance")
+
+        all_scenes = self.update_scene_record(nusc, all_samples)
+        self.logger.info("Finish updating scene")
+
+        # save
+        annotation_root = osp.join(output_path, nusc.version)
+        self._save_json(all_ego_poses, osp.join(annotation_root, "ego_pose.json"))
+        self._save_json(all_samples, osp.join(annotation_root, "sample.json"))
+        self._save_json(all_sample_data, osp.join(annotation_root, "sample_data.json"))
+        self._save_json(all_sample_anns, osp.join(annotation_root, "sample_annotation.json"))
+        self._save_json(all_instances, osp.join(annotation_root, "instance.json"))
+        self._save_json(all_scenes, osp.join(annotation_root, "scene.json"))
+
+    def interpolate_ego_pose(self, nusc: NuScenes) -> List[Dict[str, Any]]:
+        """
+        Extend ego pose records with interpolation.
+
+        The keys of ego pose are as follows.
+        * token (str)
+        * translation (list[float])
+        * rotation (list[float])
+        * timestamp (int)
+
+        Args:
+            nusc (NuScenes)
+
+        Returns:
+            List[Dict[str, Any]]
+        """
+        timestamps_msec = []
+        translations = []
+        rotations = []
+        all_ego_poses: List[Dict[str, Any]] = sorted(nusc.ego_pose, key=lambda e: e["timestamp"])
+        for ego_pose in all_ego_poses:
+            timestamps_msec.append(ego_pose["timestamp"] * 1e-3)
+            translations.append(ego_pose["translation"])
+            rotations.append(ego_pose["rotation"])
+        translations = np.array([t for _, t in sorted(zip(timestamps_msec, translations))])
+        rotations = np.array([r for _, r in sorted(zip(timestamps_msec, rotations))])
+
+        timestamps_msec, unique_idx = np.unique(timestamps_msec, return_index=True)
+        translations = translations[unique_idx]
+        rotations = rotations[unique_idx]
+        rotations = Rotation.from_quat(rotations)
+
+        func_xyz = CubicSpline(timestamps_msec, translations)
+        func_rot = Slerp(timestamps_msec, rotations)
+
+        num_times = len(timestamps_msec)
+        inter_times = np.concatenate(
+            [
+                np.arange(
+                    timestamps_msec[i] + self._interpolate_step_msec,
+                    timestamps_msec[i + 1],
```

Comment:
[Q] 
Since timestamp_msec contains all the timestamps of ego_pose, timestamp_msec[i+1] does not represent the timestamp of the next lidar but rather the timestamp of the next sensor (most likely, a camera). Is it as intended?
Is this

Reply:
[Memo] I don't have to interpolate ego pose because it is annotated for all sensors at each step.